src/main/scala/analysis/data_structure_analysis/LocalPhase.scala

```diff
@@ -310,52 +305,53 @@ class LocalPhase(proc: Procedure,
-        assert(size % 8 == 0)
-        val byteSize = size / 8
-        val global = isGlobal(index, n, byteSize)
-        val stack = isStack(index, n)
-        val indexCell = graph.adjust(graph.accessIndexToSlice(store))
-        assert(!indexCell.node.get.flags.merged) // check index cell is a placeholder
-        val addressPointee: Cell = if (global.isDefined) {
-          graph.mergeCells(graph.find(global.get), indexCell)
-          graph.adjust(graph.find(global.get).getPointee)
-        } else if (stack.isDefined) {
-          graph.mergeCells(graph.find(stack.get), indexCell)
-          graph.adjust(graph.find(stack.get).getPointee)
-        } else {
-            index match {
-              case BinaryExpr(op, arg1: Variable, arg2) if op.equals(BVADD) =>
-                evaluateExpression(arg2, constProp(n)) match {
-                  case Some(v) =>
-                    //                    assert(varToSym(n).contains(arg1))
-                    val offset = v.value
-                    visitPointerArithmeticOperation(n, indexCell, arg1, byteSize, false, offset)
-                    visitPointerArithmeticOperation(n, Node(Some(graph)).cells(0), arg1, byteSize, true, offset)
-                  case None =>
-                    //                    assert(varToSym(n).contains(arg1))
-                    // collapse the results
-                    // visitPointerArithmeticOperation(n, DSN(Some(graph)).cells(0), arg1, byteSize, true, 0, true)
-                    unsupportedPointerArithmeticOperation(n, index, Node(Some(graph)).cells(0))
-                }
-              case arg: Variable =>
-                //                assert(varToSym(n).contains(arg))
-                visitPointerArithmeticOperation(n, indexCell, arg, byteSize, false)
-                visitPointerArithmeticOperation(n, Node(Some(graph)).cells(0), arg, byteSize, true)
-              case _ =>
-                ???
-            }
-        }
         val unwrapped = unwrapPaddingAndSlicing(expr)
         unwrapped match {
           // Mem[Ry] = Rx
           case value: Variable =>
+            val index: Expr = unwrapPaddingAndSlicing(ind)
             reachingDefs(n)(value).foreach(visit)
-            addressPointee.node.get.flags.modified = true
+            assert(size % 8 == 0)
+            val byteSize = size / 8
+            val global = isGlobal(index, n, byteSize)
+            val stack = isStack(index, n, byteSize)
+            val indexCell = graph.adjust(graph.accessIndexToSlice(store))
+            assert(!indexCell.node.get.flags.merged)
+
             val valueCells = graph.getCells(n, value)
-            val result = valueCells.foldLeft(addressPointee) { (c, slice) =>
+            val valueCell = valueCells.tail.foldLeft(graph.adjust(valueCells.head)) { (c, slice) =>
               graph.mergeCells(graph.adjust(slice), c)
             }
+
+            if global.isDefined then
+              graph.mergeCells(graph.find(global.get), indexCell)
```

Comment:
Is this not doing the same exact merges with indexCell that the other structure is doing? I think the only thing that is changes is that first we merge all the possible valueCells with each other and then merging them with the addressPointee instead of getting the address pointee and then merging it with value cells

Reply:
I don't really remember why I figured out that change in order was necessary but it made a clear difference. 

https://github.com/UQ-PAC/BASIL/tree/points_to_analysis_overlapping_access_merge_test T

This branch has been merged to maintain that structure and with that all the memory region tests pass but it makes some of the tests in DataStructureAnalysisTest fail - can you have a look at that?

Reply:
Could you please take a look at [this branch](https://github.com/UQ-PAC/BASIL/tree/points_to_analysis_overlapping_access_oldorder). I have reordered the merging operations to match the old way and now all the DSAMemoryREgionSystemTestsBAP pass but the gtirb ones throw an error where accessesEndian is empty

https://github.com/UQ-PAC/BASIL/blob/e68d1a81d81bf2da74856933a4b0ac4daafdfd26/src/main/scala/translating/IRToBoogie.scala#L278

Reply:
Ah thanks, great work. I should be able to fix that, I ran into a similar bug before.

Reply:
There are memory sections getting created with size 0 because the larges accessed size of them are 0

Reply:
I remember why the change in structure was needed - the old version ignored the case where a Literal is written to a memory location. In practice this is only ever 0 I think? But it meant that cells weren't updated properly for those accesses.

Reply:
Yeah, the largestAccessedSize appears to be incorrect in those cases, but I haven't figured out why yet.

Reply:
An example is secret_write/clang:O2:GTIRB - the cell representing x doesn't have its largestAccessedSize updated and remains at 0, even though x is clearly accessed in the program. It may be because the access to x is storing 0 in it?

Reply:
https://github.com/UQ-PAC/BASIL/blob/7fe1979cb03b7c6d44cd212fd38f39caa3a26958/src/main/scala/analysis/data_structure_analysis/LocalPhase.scala#L347-L354

Is it that the largestAccessedSize of a cell only gets updated during a store if this part is reached?

Reply:
Okay I think the problem was that if a literal is assigned that DSA knows is not an address to a global, it ignored it but now it grows the memory region it was assigned to that size

Reply:
It should grow the cell but it doesn't, I think?

Reply:
I don't know why it worked for the adt version without that tho

Reply:
> It should grow the cell but it doesn't, I think?

Yeah this def was the case

Reply:
That seems to have fixed it, great! I don't know why there was a difference either, I'll give it a closer look next week to see if there's anything obvious there.